libs/prebuilt/langgraph/prebuilt/memory_handler.py

```diff
@@ -0,0 +1,279 @@
+from langchain_core.messages import AIMessage, BaseMessage, HumanMessage, SystemMessage
+from langchain_core.messages.utils import (
+    _get_message_openai_role,
+)
+from langchain_core.messages.utils import (
+    trim_messages as trim_messages_core,
+)
+from langchain_core.prompts.chat import ChatPromptTemplate, ChatPromptValue
+from langchain_core.runnables import RunnableConfig
+
+from langgraph.store.base import BaseStore
+
+SUMMARIES_NS = ("summaries",)
+
+TokenCounter = Callable[[list[BaseMessage]], int]
+
+
+DEFAULT_INITIAL_SUMMARY_PROMPT = ChatPromptTemplate.from_messages(
+    [
+        ("placeholder", "{messages}"),
+        ("user", "Create a summary of the conversation above:"),
+    ]
+)
+
+
+DEFAULT_EXISTING_SUMMARY_PROMPT = ChatPromptTemplate.from_messages(
+    [
+        ("placeholder", "{messages}"),
+        (
+            "user",
+            "This is summary of the conversation to date: {existing_summary}\n\n"
+            "Extend the summary by taking into account the new messages above:",
+        ),
+    ]
+)
+
+DEFAULT_FINAL_SUMMARY_PROMPT = ChatPromptTemplate.from_messages(
+    [
+        # if exists
+        ("placeholder", "{system_message}"),
+        ("system", "Summary of conversation earlier: {summary}"),
+        ("placeholder", "{messages}"),
+    ]
+)
+
+DEFAULT_APPROXIMATE_TOKEN_LENGTH = 4
+DEFAULT_EXTRA_TOKENS_PER_MESSAGE = 3
+
+
+def count_tokens_approximately(
+    messages: list[BaseMessage],
+    token_length: int = DEFAULT_APPROXIMATE_TOKEN_LENGTH,
+    extra_tokens_per_message: int = DEFAULT_EXTRA_TOKENS_PER_MESSAGE,
+    include_name: bool = True,
+) -> int:
+    token_count = 0
+    for message in messages:
+        message_chars = 0
+        if isinstance(message.content, str):
+            message_chars += len(message.content)
+
+        # TODO: handle image content blocks properly
+        else:
+            content = json.dumps(message.content)
+            message_chars += len(content)
+
+        if (
+            isinstance(message, AIMessage)
+            # exclude Anthropic format as tool calls are already included in the content
+            and not isinstance(message.content, list)
+            and message.tool_calls
+        ):
+            tool_calls_content = json.dumps(message.tool_calls)
+            message_chars += len(tool_calls_content)
+
+        role = _get_message_openai_role(message)
+        message_chars += len(role)
+
+        if message.name and include_name:
+            message_chars += len(message.name)
+
+        # NOTE: we're rounding up per message to ensure that the token counts
+        # are always consistent
+        token_count += math.ceil(message_chars / token_length)
+
+        # add extra tokens per message
+        # see this https://github.com/openai/openai-cookbook/blob/main/examples/How_to_count_tokens_with_tiktoken.ipynb
+        token_count += extra_tokens_per_message
+
+    return token_count
+
+
+def trim_messages(
```

Comment:
This is going to be really confusing; e.g., imagine what happens when users search for trim_messages on google